sdstore/md5helper.go

```diff
@@ -0,0 +1,94 @@
+// Taken and modified from https://blog.golang.org/pipelines
+
+package sdstore
+
+import (
+	"crypto/md5"
+	"encoding/hex"
+	"errors"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sync"
+)
+
+// A result is the product of reading and summing a file using MD5.
+type result struct {
+	path string
+	// sum  [md5.Size]byte
+	sum string
+	err error
+}
+
+// sumFiles starts goroutines to walk the directory tree at root and digest each
+// regular file.  These goroutines send the results of the digests on the result
+// channel and send the result of the walk on the error channel.  If done is
+// closed, sumFiles abandons its work.
+func sumFiles(done <-chan struct{}, root string) (<-chan result, <-chan error) {
+	// For each regular file, start a goroutine that sums the file and sends
+	// the result on c.  Send the result of the walk on errc.
+	c := make(chan result)
+	errc := make(chan error, 1)
+	go func() { // HL
+		var wg sync.WaitGroup
+		err := filepath.Walk(root, func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+			if !info.Mode().IsRegular() {
+				return nil
+			}
+			wg.Add(1)
+			go func() { // HL
+				data, err := ioutil.ReadFile(path)
```

Comment:
Wonder how efficient this is.  
Read all files and calculate md5 vs just blindly upload the entire zip... 
For something like `node_modules`, this will take a lot of time.

Reply:
Yep, for any sort of `md5`ing we're going to have to open all the contents, but I guess the tradeoff is whether we have smaller caches in general or bigger ones. If the usual cache is something like `node_modules/` this might not be worth having